src/OrleansCodeGenerator/RoslynCodeGenerator.cs

```diff
@@ -429,10 +429,13 @@ private static GeneratedSyntax GenerateForAssemblies(List<Assembly> assemblies,
             if (typeInfo.IsSerializable)
                 RecordType(type, module, targetAssembly, includedTypes);
 
-            Type grainStateType;
+            Type typeToRecord;
             // If a type extends Grain<T>, add T to the list of candidates for serialization
-            if (IsPersistentGrain(typeInfo, out grainStateType))
-                RecordType(grainStateType, module, targetAssembly, includedTypes);
+            if (IsPersistentGrain(typeInfo, out typeToRecord))
+                RecordType(typeToRecord, module, targetAssembly, includedTypes);
+
+            if (IsAsyncObserver(typeInfo, out typeToRecord))
+                RecordType(typeToRecord, module, targetAssembly, includedTypes);
```

Comment:
Record really just means "record if serializable", it performs accessibility checks first